### DISCUSSION TITLE
multi workspace

### DISCUSSION THREAD

**Discussion by PirroFrIta** [usememos/memos#4155] [Ideas]
It would be great to have the ability to create multiple workspaces with a table to manage various member logins

|     | Workspace 1 | Workspace 2 | Workspace 3 |
| --------- | ----------- | ----------- | ----------- |
| user 1   |        |        |        |
| user 2 |      |        |        |
| user 3 |   x   |    x    |        |
| user 4 |      |        |        |
| user 5 |      |    x    |        |

**Comment by ben-towner** [comment]
I agree, multiple workspaces would be ideal for splitting the memos into work, home, different teams, etc.

**Comment by Dewep** [comment]
Same feelings here, I am using memos for multiple projects (home, main job, side-projects, technical notes, ...), and I would love being able to better split them.

This is kind of a more advanced feature, because I would say your "user 3", in 2 different workspace, should be able to have 2 lists of memos. Having one list, with both workspace content would be kind of messy I think?

**Reply by mdg-pnw** [reply]
> This is kind of a more advanced feature, because I would say your "user 3", in 2 different workspace, should be able to have 2 lists of memos. Having one list, with both workspace content would be kind of messy I think?

I'd imagine that Workspaces would looks something like this:
![image](https://github.com/user-attachments/assets/23b63d9f-aea9-4e81-9265-a18aea5b30de)


Home would still show you all of the memos that you have access to in one combined view. Clicking on a workspace would just show you memos that are just in that workspace.

**Comment by zachpro1935** [comment]
bump... Need this feature...

**Comment by Novapixel1010** [comment]
I really need this feature. I wish I knew more code I wonder how much it would cost to make this happen?

**Comment by Novapixel1010** [comment]
Here is my fully length proposal for this feature

# Why add multi-workspaces?

Today Memos supports per-memo visibility (private / protected / public) but “protected” effectively means everyone on the instance. That makes it hard to share a subset of notes with just a project team or family group. 

## Benefits

- Least-privilege sharing: Users can share with some members without exposing to the whole server (“protected”).

- Cleaner organization: One account, many contexts (personal, work, family extended, Everyone in the same house).

- Enterprise-readiness: Paves the way for role-based access control (RBAC), audit, and admin tooling per workspace.

- Backwards compatible path: You can migrate current “protected” visibility into a default workspace with no behavior change.

## Core concept

A **Workspace** represents a collaboration boundary: a set of users and the memos they can see. A memo can be:

- PRIVATE (author only)

- WORKSPACE (visible to members of its workspace)

- PUBLIC (world)

This is the same visibility model users already know, but “protected” is now workspace-scoped instead of instance-wide.

## Data model (DB schema)

Add three tables (works across SQLite/MySQL/Postgres, all currently supported by

1. workspace

- id (PK)

- name, slug (unique), description

- created_ts, updated_ts

- created_by

2. workspace_member

- workspace_id 

- user_id

- Composite PK (workspace_id, user_id)

3. Option A (simplest) – one workspace per memo: add workspace_id  to memo

If NULL and visibility=PRIVATE, treat as personal memo.

If visibility=WORKSPACE, workspace_id must be non-null.

(Option B – many workspaces per memo via memo_workspace join table. Start with A; you can evolve to B if/when “share same memo across multiple workspaces” is requested.)

### Migration plan

- Create a built-in **“Default”** workspace.

- For each existing memo with “visible to members,” set visibility=WORKSPACE and workspace_id=<Default>.

- Keep PRIVATE and PUBLIC memos unchanged.

### Permissions / RBAC rules

- Read memo:

     - PUBLIC: anyone.

     - WORKSPACE: requester = workspace_member.

     - PRIVATE: author only.

- Write memo: author of the memo’s.

- Manage workspace:

     - OWNER full control; ADMIN can manager you can create a workspace and members in that workspace.

### API surface (v1/v2)

Add endpoints (REST) alongside existing APIs

#### Workspaces

- POST /api/workspaces – create

- GET /api/workspaces – list mine

- GET /api/workspaces/:id – get

- PATCH /api/workspaces/:id – rename/desc

- DELETE /api/workspaces/:id – delete (soft-delete first)

#### Members

- GET /api/workspaces/:id/members

- PUT /api/workspaces/:id/members/:userId – upsert with role

- DELETE /api/workspaces/:id/members/:userId

#### Memos

- Extend existing memo create/update payloads with workspaceId and visibility.

- Enforce workspaceId required IF visibility=WORKSPACE.

- Query filters: ?workspaceId=…, plus existing tag/shortcut filters

#### OpenAPI / SDKs

- Version new fields; default workspaceId=null for backwards compatibility.

### UI/UX changes (web app)

- Workspace switcher in the global header (with search + quick jump).

- Composer: a two-step share control:

    - Visibility dropdown: Private / Workspace / Public

     - If Workspace, show a workspace selector (defaults to “last used” for convenience).

- Sidebar: new “Workspaces” section.

- Settings have toggle to turn on workspaces basically just toggling its visibility I assume this will make the code less complex.
     - Tab called "workspaces"
        - In there you can create a workspace and add members to it. Maybe also select which members are allowed to create a workspace.   
        - Also control things per workspace like toggling comments on or off. 

- Explore: filter by workspace; keep public explore as is.

- Shortcuts / filters: allow `workspaceId == 123`.

### What needs to change in the codebase ?

I am not sure I am assume the database,frontend,server.

### Backward compatibility & rollout

1. Ship with feature flag ON but auto-creates “Default” workspace and auto-assigns all existing “member-visible” memos to it; behavior matches today.

2. Admin can create new workspaces and move memos (batch actions are a frequent UX request). 

### Edge cases & future work

- Cross-workspace mentions / links: allow linking by URL; show a lock badge if the viewer lacks access.

- Multi-workspace memos (Option B) if users want the same memo shared to multiple groups.

- Per-workspace integrations (webhooks, export).

- Audit log per workspace for enterprise use.

- Collaborative editing can be layered on top of workspace scoping


So what do you think @boojack

**Reply by 1joneselliot** [reply]
Multi-workspaces would be great, but at a minimum, per-note sharing to individual users is a pretty essential evolution to the existing sharing model

**Comment by DarkRanger935** [comment]
Yes please.  This hit the nail on the head for one of the most basic things to really evolve past "super useful for an individual but not so much for an organization"

**Comment by boojack** [comment]
Thanks everyone for the thoughtful proposals and patience. Good news—this is now available for testing as **Spaces** in [Memos v0.31.0-rc.1](https://github.com/usememos/memos/releases/tag/v0.31.0-rc.1).

You can create multiple Spaces, invite existing instance users as `ADMIN` or `USER`, and switch between the combined **Memos** view and an individual Space. Selecting a Space scopes Home, Explore, memo creation, statistics, filters, and attachments to that Space.

Each memo can remain **Unassigned** or belong to exactly one Space while staying owned by its author. Set its visibility to **Space** when it should be readable only by that Space’s members; **Protected** keeps its existing instance-wide meaning.

Since this is a release candidate, we’d appreciate your testing and feedback—especially around membership, switching between contexts, and upgrade behavior.

**Reply by sjorge** [reply]
Is there a way to view only 'unassigned' memos too?

Say I have 10 unassigned memos and 10 in a space, if I hit the top option I get all 20, if I pick the space I get the 10 from the space.

There seems to be no way to get only the 10 unassigned (personal) memos, nor does there seem to be a way to set a space (as the default view or well the unassigned view that doesn't exist 😂 )

I guess an option to hide memos belonging to a space for the default view would also do the trick.

**Reply by Novapixel1010** [reply]
Testing it now not sure if you want to open a second discussion for talking about spaces so I will just leave a comment here. So far it looks amazing, before I suggest and features I want to say I think the focus should be making sure it's **stable** and works good. 

## Some things I noticed 
### public page
I noticed when going to `https://insert-domain.com/explore` it shows all public notes and think it would be better to do something like `https://insert-domain.com/space1/explore` and the default `https://insert-domain.com/explore` being only the public memos from the main/default space. It would also be nice in the space settings under `General` to disable the ability for that space to have public option. 
### compose file change
This not really a big issue just noticed it. 
My current compose file has `MEMOS_INSTANCE_URL: insert-domain.com` this env no longer works and I an error in the logs says must be http or https. So I changed it to `MEMOS_INSTANCE_URL: https://insert-domain.com` and the container worked.
## Later features/fixes
### user space stuff
- invite users to space **Update**: This is added and works #6266 
~~It would be great if when creating a user right under role. You could at least invite them to spaces, what would make this even better is if the user could set there default space that they would like after login. A invite should be sent to the user inbox to join the space, currently the user doesn't know they have an invite unless they go to there settings. (Also I would like to note have not setup email yet, I will later). I talking about the inbox in memos not email.~~
- view options
I love that I can view all my memos in one spot, but once you start having a lot of them it would be great in the `View Options` to select which spaces you want to be able to see/view. 
- create space
In system setting there should be a toggle to disallow non-admin users from creating a space.
### space theme
As a later feature to work on it would be cool to set a default theme to use for that space. 

---
I'll add more if I come across anything else. Just a heads-up, I'm currently on version `0.27.1` and I'm holding off on upgrading until this feature is wrapped up. I'm happy to test out other things, so if you need anything specific checked, just let me know.

**Reply by sjorge** [reply]
There is currently no way (via webui) at least to move a memo unassigned and being assigned to a space, perhaps there could be a dropdown added to the menu when clicking the date in edit mode to select an optional space for the memo.

Edit: this is now possible! Yay.

I also found a way to set a "default" space when loading, I create 2 spaces: personal, and work, I want to default to personal. I managed to do this with mod_rewrite on the proxy since there is now space routes expose via the URL.

```
RewriteEngine On
RewriteRule ^/$ /spaces/<space_id> [R=302,L]
```

This works and auth redirect and such all keeps working (as far as I can tell) of course being able to pick a 'default' space in settings would be better but this is good enough for me at the moment as I am the only one using my instance.

**Reply by Novapixel1010** [reply]
>There is currently no way (via webui) at least to move a memo unassigned

This has been added and looks good.

<img width="394" height="392" alt="move-space" src="https://github.com/user-attachments/assets/d9df7e3b-2b47-4c2c-a8fa-a03b46c3d1c6" />

I did notice something a little weird maybe a **BUG**. But you can also move comments

<img width="763" height="408" alt="funny-spaces-thing" src="https://github.com/user-attachments/assets/3b15159e-0b6e-4c3c-8175-e8995fe5e248" />

**Reply by boojack** [reply]
**Updates**: [Memos v0.31.0-rc.2](https://github.com/usememos/memos/releases/tag/v0.31.0-rc.2) is now available, with several Spaces updates based on the feedback here:

- **Move memos:** Use **Move to Space** in a memo’s menu to move your memos between Spaces or back to **Unassigned**.
- **Find unassigned memos:** Create a saved View with `space == null` to see only memos outside any Space.
- **Inbox invitations:** Space invitations now appear in the recipient’s inbox.
- **Direct Space links:** Each Space has its own URL. You can bookmark it, and switching Spaces preserves your current page and filters.
- **Custom icons:** Choose icons or emoji for Spaces, and custom icons for saved Views.

Thanks for testing and sharing your feedback!

**Comment by boojack** [comment]
Quick update: **Spaces is now available in the stable [Memos v0.31.0 release](https://github.com/usememos/memos/releases/tag/v0.31.0)!** You can create multiple Spaces, invite members, and organize memos by context. Thanks everyone for the ideas, testing, and feedback that helped shape this feature!
